### PR TITLE
Add HostAlignedByteCount to enforce alignment at compile time

### PR DESCRIPTION
**Pull request by sunshowers**
As part of the work to allow mmaps to be backed by other implementations (#9544), I realized that we didn't have any way to track whether a particular usize is host-page-aligned at compile time.

Add a `HostAlignedByteCount` which tracks that a particular usize is aligned to the host page size. This also does not expose safe unchecked arithmetic operations, to ensure that overflows are always handled.

With `HostAlignedByteCount`, a lot of runtime checks can go away thanks to the type-level assertion.

In the interest of keeping the diff relatively small, I haven't converted everything over yet. More code can be converted over as we go along.

### CODE REVIEW

File: crates/wasmtime/src/runtime/vm/instance/allocator/pooling/unix_stack_pool.rs [hunk 1]
```diff
@@ -152,11 +158,11 @@ impl StackPool {
...
 
         // Remove the guard page from the size
-        let stack_size = self.stack_size - self.page_size;
+        let stack_size = self.stack_size.byte_count() - self.page_size.byte_count();
         let bottom_of_stack = top - stack_size;
-        let start_of_stack = bottom_of_stack - self.page_size;
+        let start_of_stack = bottom_of_stack - self.page_size.byte_count();
         assert!(start_of_stack >= base && start_of_stack < (base + len));
-        assert!((start_of_stack - base) % self.stack_size == 0);
+        assert!((start_of_stack - base) % self.stack_size.byte_count() == 0);
```
Comment (sunshowers) [inline crates/wasmtime/src/runtime/vm/instance/allocator/pooling/unix_stack_pool.rs:165]: This code and the similar section below would be good to convert over, but I decided to do this in a followup.

File: crates/wasmtime/src/runtime/vm.rs [hunk 2]
```diff
@@ -336,6 +336,216 @@ impl ModuleRuntimeInfo {
     }
 }
 
+/// A number of bytes that's guaranteed to be aligned to the host page size.
+///
+/// This is used to manage page-aligned memory allocations.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct HostAlignedByteCount(
```
Comment (sunshowers) [inline crates/wasmtime/src/runtime/vm.rs:343]: I first named this `AlignedByteCount`, but then realized that the VM page size might not be the same as the host page size, so named it `HostAlignedByteCount` instead. Happy to rename this to something else if it makes sense.

File: crates/wasmtime/src/runtime/vm.rs [hunk 3]
```diff
@@ -336,6 +336,216 @@ impl ModuleRuntimeInfo {
...
+impl PartialEq<HostAlignedByteCount> for usize {
+    #[inline]
+    fn eq(&self, other: &HostAlignedByteCount) -> bool {
+        *self == other.0
+    }
+}
+
+impl fmt::Display for HostAlignedByteCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
```
Comment (sunshowers) [inline crates/wasmtime/src/runtime/vm.rs:483]: Should this print the size in hex or as a decimal?

**Comment by sunshowers** [comment]
I started working on a followup as well: https://github.com/bytecodealliance/wasmtime/commit/03efe42fb4d8fdd760d945ce6ceb95b8ca338974

Unfortunately GitHub's review flow does not let you manage stacked PRs unless you have push access to the destination repo, so I'll have to put these up one at a time. Sigh :(

File: crates/wasmtime/src/runtime/vm.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/wasmtime/src/runtime/vm.rs:343]: Same seems reasonable to me 👍

File: crates/wasmtime/src/runtime/vm.rs [hunk 4]
```diff
@@ -336,6 +336,216 @@ impl ModuleRuntimeInfo {
...
+/// A number of bytes that's guaranteed to be aligned to the host page size.
+///
+/// This is used to manage page-aligned memory allocations.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct HostAlignedByteCount(
+    // Invariant: this is always a multiple of the host page size.
+    usize,
+);
+
+impl HostAlignedByteCount {
```
Comment (alexcrichton) [inline crates/wasmtime/src/runtime/vm.rs:348]: Mind putting this type/impls/errors/etc in a dedicated submodule of the `vm` module?

File: crates/wasmtime/src/runtime/vm.rs [hunk 5]
```diff
@@ -336,6 +336,216 @@ impl ModuleRuntimeInfo {
...
+#[cfg(feature = "std")]
+impl std::error::Error for ByteCountOutOfBounds {}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum ByteCountOutOfBoundsKind {
+    RoundUp(usize),
+    U64(u64),
+    Add(usize, usize),
+    Sub(usize, usize),
+    Mul(usize, usize),
```
Comment (alexcrichton) [inline crates/wasmtime/src/runtime/vm.rs:524]: For here I think it's probably ok to discard the precise values that were being operated on as otherwise that makes the errors here pretty large for a low-level type such as this.

File: crates/wasmtime/src/runtime/vm.rs (hunk 3 above)
Comment (alexcrichton) [reply crates/wasmtime/src/runtime/vm.rs:483]: Let's go with hex perhaps to clearly show that the low bits are all zero

**Review by alexcrichton — APPROVED** [review]
Thanks for this!

File: crates/wasmtime/src/runtime/vm/sys/unix/mmap.rs [hunk 6]
```diff
@@ -134,8 +138,10 @@ impl Mmap {
     }
 
     #[inline]
-    pub fn len(&self) -> usize {
-        self.memory.as_ptr().len()
+    pub fn len(&self) -> HostAlignedByteCount {
+        // SAFETY: When self.memory was created it was always aligned to a page
+        // boundary.
+        unsafe { HostAlignedByteCount::new_unchecked(self.memory.as_ptr().len()) }
```
Comment (alexcrichton) [inline crates/wasmtime/src/runtime/vm/sys/unix/mmap.rs:144]: Reviewing some other code elsewhere, I think that the `from_file` constructor may cause problems here because file lengths aren't always page-aligned. 

To help catch this mind adding a `debug_assert!` to `new_unchecked`? That I think should make the test suite panic

**Comment by alexcrichton** [comment]
As a heads up I'm adding https://github.com/bytecodealliance/wasmtime/pull/9614 to the merge queue which is going to have a number of conflicts with this. It should (hopefully) be just a few `#[cfg(feature = "signals-based-traps")]` in a few places but if anything is overly difficult or onerous lemme know and I can help out

File: crates/wasmtime/src/runtime/vm.rs (hunk 4 above)
Comment (sunshowers) [reply crates/wasmtime/src/runtime/vm.rs]: Done.

File: crates/wasmtime/src/runtime/vm/sys/unix/mmap.rs (hunk 6 above)
Comment (sunshowers) [reply crates/wasmtime/src/runtime/vm/sys/unix/mmap.rs]: This ended up becoming #9639.

File: crates/wasmtime/src/runtime/vm.rs (hunk 5 above)
Comment (sunshowers) [reply crates/wasmtime/src/runtime/vm.rs]: Hmm good point, done.

**Comment by sunshowers** [comment]
(Apologies for the force push -- had to do it as part of the rebase. GH makes me cry)

**Review by alexcrichton — APPROVED** [review]
Thanks again for this!

**Comment by sunshowers** [comment]
All right, think I fixed the Miri issue.

File: crates/wasmtime/src/runtime/vm/memory/mmap.rs [hunk 7]
```diff
@@ -178,23 +190,20 @@ impl RuntimeLinearMemory for MmapMemory {
...
-            assert!(
-                new_accessible <= self.mmap.len() - self.offset_guard_size - self.pre_guard_size,
-            );
 
             // If the Wasm memory's page size is smaller than the host's page
             // size, then we might not need to actually change permissions,
             // since we are forced to round our accessible range up to the
             // host's page size.
-            if new_accessible > self.accessible() {
+            if let Ok(difference) = new_accessible.checked_sub(self.accessible()) {
```
Comment (alexcrichton) [inline crates/wasmtime/src/runtime/vm/memory/mmap.rs:200]: I think this may be the source of the bug, albeit it's a bit indirect. Previously the `>` condition is now morally becoming `>=` since `new_accessible == self.accessible()` didn't run this block before but it's now running this block with `difference == 0`. I think then it may be the case that mprotect for 0 bytes is succeeding on unix but failing on Windows.

On Linux I see `mprotect(0x740d36001000, 0, PROT_READ|PROT_WRITE) = 0` in strace with this patch when running the failing spec test and while I haven't executed this on Windows this is what I would suspect is going on.

Another option is to update the `Mmap` type to skip `make_accessible` when the byte size is zero which I think would also be reasonable.

Comment (sunshowers) [reply crates/wasmtime/src/runtime/vm/memory/mmap.rs:200]: Oh that's terrifying -- I'm going to fix this (probably by making `make_accessible` ignore zero-sized regions) and go through this code again to make sure I didn't miss any other sign changes. Thanks for diagnosing this!

Comment (sunshowers) [reply crates/wasmtime/src/runtime/vm/memory/mmap.rs:200]: Made `make_accessible` a no-op on zero-length allocations, as well as the other mprotect calls (`make_executable` and `make_readonly`). They also had this as a lurking bug because they had assertions that `range.start <= range.end`, not `range.start < range.end`.

I've also added tests for zero-length calls, and re-reviewed everything -- didn't see any other changes.
